added cpp and header file for new optimization subsolver

mcjuneho opened the pull request:
Wanted to make a draft PR without actually adding it as you suggested earlier, but didn't want to accidentally make a real PR to cvc4. 

This master is up to date with cvc4 master and so this PR should reflect what the real PR to cvc4 master should look like.

Commit:
added cpp and header file for new optimization subsolver

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,108 @@
+    {
+      optChecker->push();
+      Objective o = d_activatedObjectives[i];
+      // CVC4::Kind k = o.d_node.getKind();
```
needed?

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,108 @@
+
+      while (loop_r.isSat())
+      {
+        Node value = optChecker->getValue(o.d_node);
```
`Node value;` should be declared outside the loop without assignment.

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,108 @@
+      {
+        Node value = optChecker->getValue(o.d_node);
+        o.d_savedValue = value;
+        Node increment;
```
`Node increment;` should be declared outside the loop without assignment.

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,108 @@
+}
+
+}  // namespace smt
+}  // namespace CVC4
```
Please end the file with a newline (maybe `clang-format` will do this automatically?)

yoni206 reviewed (changes requested):
Thanks for this PR!
I wrote some specific comments inline. In addition, here are some more general comments:

1. Can you include a unit test already in this PR? 
2. Please add documentation to the header file: each member and function should be documented, preferably using `@param` and `@return`. A good example for that is [here](https://github.com/CVC4/CVC4/blob/6235612ddfdd59432eedf771a65c248d5b3d0469/src/smt/preprocessor.h#L74). Note that for simple functions it is ok to just describe what they do shortly (there are examples for that too in the same link).
3. Please add documentation to the cpp file, inside the implementation of each function. It doesn't have to be before every line, but perhaps before every block and/or around lines that are not self-explanatory.

ahmed-irfan commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,108 @@
+/*! \file optimization_solver.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Ying Sheng
```
update author

ahmed-irfan commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,108 @@
+    optChecker->setOption("incremental", "true");
+    optChecker->setOption("produce-models", "true");
+
+    for (const Node& e : axioms){
```
format

ahmed-irfan commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,108 @@
+
+}
+
+void OptimizationSolver::activateObj(const Node& obj, const int& type, const int& result){
```
clang-format

ahmed-irfan commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,94 @@
+/*! \file optimization_solver.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Ying Sheng
```
update author

ahmed-irfan reviewed (changes requested):
minor formatting changes

Commit:
added unit tests and made small format fixes

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,94 @@
+}  // namespace smt
+}  // namespace CVC4
+
+#endif /* CVC4__SMT__OPTIMIZATION_SOLVER_H */
```
please end all files with a newline.

yoni206 commented on test/unit/theory/theory_int_opt_white.h:
```diff
@@ -0,0 +1,78 @@
+ public:
+  void setUp() override
+  {
+    //d_optslv(nullptr)
```
please remove commented out code.

yoni206 commented on test/unit/theory/theory_int_opt_white.h:
```diff
@@ -0,0 +1,78 @@
+  }
+  void testOmt()
+  {
+    Term ub = d_slv->mkReal(100);
```
Please add documentation: what formula are you creating? what is the expected solution?
It is best to add some comments before each block of code to explain what it does.

yoni206 commented on test/unit/theory/theory_int_opt_white.h:
```diff
@@ -0,0 +1,78 @@
+    CVC4::Result r;
+    d_optslv->checkOpt(r);
+
+    std::cout << "Result is :" << r << std::endl;
```
printing is great, but please also add some assertions to make sure that the result is the expected one. 
See, e.g.,  `EXPECT_EQ` statements 
like here:
https://github.com/mcjuneho/CVC4/blob/a54d5e4192071e6fe85d21e07a9091ff7ccd08c2/test/unit/theory/evaluator_white.h#L174

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,109 @@
+    optChecker->assertFormula(e);
+  }
+
+  NodeManager* nm = optChecker->getNodeManager();
```
Please document the code with comments, explaining what it does.

yoni206 reviewed (commented):
Looks good, thanks for this. I think we are close to being ready to submit a PR to the main repo. I have some comments below, please try to address them first.

Commit:
added documentation and small style fixes

Commit:
ran git clang-format

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,128 @@
+    Node value;
+    Node increment;
+
+    /*Workhorse of linear optimization:
```
start every line with //

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,94 @@
+    OPT_OPTIMAL
+  };
+
+  enum CVC4_PUBLIC ObjectiveType
```
add doc on all this enum also

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,94 @@
+    OBJECTIVE_MAXIMIZE
+  };
+
+  enum CVC4_PUBLIC ObjectiveValue
```
add doc to enum

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,94 @@
+  {
+   public:
+    Objective(Node n, ObjectiveType type, OptResult result);
+    //~Objective();
```
```suggestion
    ~Objective() {}
```

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,94 @@
+    Objective(Node n, ObjectiveType type, OptResult result);
+    //~Objective();
+
+    /*private:*/
```
please document each field. 
/* .... */

/**
 *
 *
 */

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,94 @@
+  OptimizationSolver(SmtEngine* parent);
+  ~OptimizationSolver();
+
+  bool checkOpt(Result& r);
```
document these functions

yoni206 commented on test/unit/CMakeLists.txt:
```diff
@@ -170,4 +170,4 @@ add_subdirectory(printer)
 add_subdirectory(theory)
 add_subdirectory(preprocessing)
-add_subdirectory(util)
+add_subdirectory(util)
```
end file with \n

yoni206 commented on test/unit/theory/CMakeLists.txt:
```diff
@@ -29,3 +29,4 @@ cvc4_add_cxx_unit_test_white(theory_strings_skolem_cache_black theory)
 cvc4_add_cxx_unit_test_white(theory_strings_word_white theory)
 cvc4_add_cxx_unit_test_white(theory_white theory)
 cvc4_add_cxx_unit_test_white(type_enumerator_white theory)
+cvc4_add_cxx_unit_test_white(theory_int_opt_white theory)
```
end file witn \n

yoni206 commented on test/unit/theory/theory_int_opt_white.h:
```diff
@@ -0,0 +1,94 @@
+    std::cout << "Optimized min value is: "
+              << d_optslv->objectiveGetValue(*min_cost.d_node) << std::endl;
+  }
+};
```
end file with \n

Commit:
expanded unit test and modified subsolver to support one objective

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,114 @@
+  class OptimizationSolver
+  {
+   public:
+    OptimizationSolver(SmtEngine* parent);
```
say that parent is a subsolver that....

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,127 @@
+namespace CVC4 {
+namespace smt {
+
+OptimizationSolver::OptimizationSolver(SmtEngine* parent)
```
Should be documented:
1. say that we initialize the activated objective to a default one.
2. say that we initialize the saved value to a default node (null node)

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,127 @@
+
+OptResult OptimizationSolver::checkOpt()
+{
+  Assert(d_activatedObjective.getType() != OBJECTIVE_UNDEFINED);
```
```suggestion
  // Make sure that the objective is not the default one 
  Assert(d_activatedObjective.getType() != OBJECTIVE_UNDEFINED);
```

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,127 @@
+    return OPT_UNSAT;
+  }
+
+  Node value;
```
say what is value, what is increment.

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,114 @@
+/**
+ * An enum for optimization queries.
+ *
+ * Represenets whether an objective should be minimized: 0
```
don't use 0 or one, but the actual names.

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,114 @@
+/**
+ * An enum for optimization queries.
+ *
+ * Represents the result of a checkopt query
+ * OPT_UNKNOWN: if the original set of assertions has result UNKNOWN
+ * OPT_UNSAT: if the original set of assertions has result UNSAT
+ * OPT_SAT_PARTIAL: if the optimized value is only partial (unimplemented)
+ * OPT_SAT_APPROX: if the solver halted early and value is only approximate
+ * (unimplemented) OPT_OPTIMAL: if the optimization loop finished and optimal
+ * value was found
+ */
+enum CVC4_PUBLIC OptResult
+{
+  OPT_UNKNOWN,
+  OPT_UNSAT,
+  OPT_SAT_PARTIAL,
+  OPT_SAT_APPROX,
+  OPT_OPTIMAL
+};
```
```suggestion
 * Represents the result of a checkopt query




 * (unimplemented) OPT_OPTIMAL: if 
 * value was found
 */
enum CVC4_PUBLIC OptResult
{
  // the original set of assertions has result UNKNOWN
  OPT_UNKNOWN,
  // the original set of assertions has result UNSAT
  OPT_UNSAT,
  // the optimization loop finished and optimal
  OPT_OPTIMAL
  // The last two values are here as a preparation for future work
  // in which partial and approximate optimizations will be supported.

   // if the optimized value is only partial
  OPT_SAT_PARTIAL,
  // if the solver halted early and value is only approximate
  OPT_SAT_APPROX,

};
```

yoni206 commented on test/unit/theory/theory_int_opt_white.h:
```diff
@@ -0,0 +1,142 @@
+  }
+  void testOmtMax()
+  {
+    Term ub = d_slv->mkReal(100);
+    Term lb = d_slv->mkReal(0);
+
+    // Objectives to be optimized max_cost is max objective
+    Term max_cost = d_slv->mkConst(d_slv->getIntegerSort(), "cost");
+
+    Term upb = d_slv->mkTerm(CVC4::api::Kind::GT, ub, max_cost);
+    Term lowb = d_slv->mkTerm(CVC4::api::Kind::GT, max_cost, lb);
+
+    /* Result of asserts is:
+       0 < max_cost < 100
+    */
+    d_slv->assertFormula(upb);
+    d_slv->assertFormula(lowb);
+
+    // max_type is 1, signifies max_objective
```
use actual enum

yoni206 commented on test/unit/theory/theory_int_opt_white.h:
```diff
@@ -0,0 +1,142 @@
+    d_slv->assertFormula(lowb);
+
+    // max_type is 1, signifies max_objective
+    const int max_type = 1;
```
also, use actual enum

yoni206 commented on test/unit/theory/theory_int_opt_white.h:
```diff
@@ -0,0 +1,142 @@
+    d_slv->assertFormula(upb2);
+    d_slv->assertFormula(lowb2);
+
+    // min_type is 0, signifies min_objective
+    const int min_type = 0;
```
use actual enum

yoni206 commented on test/unit/theory/theory_int_opt_white.h:
```diff
@@ -0,0 +1,142 @@
+    d_slv->assertFormula(upb2);
+    d_slv->assertFormula(lowb2);
+
+    // min_type is 0, signifies min_objective
+    const int min_type = 0;
```
use actual enum

mcjuneho committed:
Update src/smt/optimization_solver.h

Co-authored-by: yoni206 <yoni206@users.noreply.github.com>

Commit:
small change

Commit:
Merge branch 'optimization_subsolver' of https://github.com/mcjuneho/CVC4 into optimization_subsolver

Commit:
fixed style

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,137 @@
+  Assert(d_activatedObjective.getType() != OBJECTIVE_UNDEFINED);
+  Assert(!d_activatedObjective.getNode().isNull());
+
+  std::unique_ptr<SmtEngine> optChecker;
```
```suggestion
  // the smt engine to which we send intermediate queries
  // for the linear search.
  std::unique_ptr<SmtEngine> optChecker;
```

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,137 @@
+
+  // We need to checksat once before the optimization loop so we have a
+  // baseline value to increment
+
```
```suggestion
```

yoni206 commented on src/smt/optimization_solver.cpp:
```diff
@@ -0,0 +1,137 @@
+    // increment on the model-value of objective:
+    // if we're maximizing increment = objective > old_objective value
+    // if we're minimizing increment = objective < old_objective value
+
```
```suggestion
```

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,123 @@
+ * An enum for optimization queries.
+ *
+ * Represents the result of a checkopt query
+
+
+
+
```
```suggestion
```

yoni206 commented on src/smt/optimization_solver.h:
```diff
@@ -0,0 +1,123 @@
+  OPT_UNSAT,
+  // the optimization loop finished and optimal
+  OPT_OPTIMAL
+  // The last two values are here as a preparation for future work
```
```suggestion

    // The last two values are here as a preparation for future work
```

yoni206 reviewed (changes requested):
plz

mcjuneho committed:
Update src/smt/optimization_solver.cpp

Co-authored-by: yoni206 <yoni206@users.noreply.github.com>

mcjuneho committed:
Update src/smt/optimization_solver.cpp

Co-authored-by: yoni206 <yoni206@users.noreply.github.com>

mcjuneho committed:
Update src/smt/optimization_solver.cpp

Co-authored-by: yoni206 <yoni206@users.noreply.github.com>

mcjuneho committed:
Update src/smt/optimization_solver.h

Co-authored-by: yoni206 <yoni206@users.noreply.github.com>

mcjuneho committed:
Update src/smt/optimization_solver.h

Co-authored-by: yoni206 <yoni206@users.noreply.github.com>

Commit:
change

Commit:
Merge branch 'master' into optimization_subsolver

Commit:
fixed format issue

ajreynol committed:
Remove quantifiers regression from decision folder (#5830)

This is a duplicate of https://github.com/CVC4/CVC4/blob/master/test/regress/regress1/quantifiers/symmetric_unsat_7.smt2 and moreover is slow on proof-new.

ajreynol committed:
(proof-new) Refactor theory preprocessing (#5835)

This simplifies our algorithm for theory preprocessing. The motivation is two-fold:
(1) Proofs are currently incorrect in rare cases due to incorrectly tracking pre vs post rewrites.
(2) We can avoid traversing Boolean connectives with term formula removal. Note that this PR relies on the observation that term formula removal leaves Boolean structure unchanged and can apply locally to theory atoms.

Furthermore, notice that our handling of lemmas generated by term formula removal is now more uniform. In the second algorithm, term formula removal is not applied to fixed point anymore. In other words, we do not remove skolems from lemmas generated by term formula removal recursively. Instead, since lemmas are theory-preprocessed recursively, we use the fixed point that runs outside the above algorithm instead of relying on the inner fixed point in term formula removal.

This PR resolves several open issues with proof production on proof-new.

I will performance test this change on SMT-LIB.

ajreynol committed:
(proof-new) Miscellaneous fixes and regressions (#5841)

HanielB committed:
[proof-new] Fix MACRO_RESOLUTION expansion for singleton clause corner case (#5850)

Evaluating the proof infrastructure in SMT-LIB has uncovered a rare case (i.e., not previously in our regressions!!) in which the resulting clause from crowding literal elimination is a singleton. This commit makes the expansion code robust to that and adds an example of a problematic benchmark as a regression.

Also improves a bit tracing and some comments.

mpreiner committed:
Add BV solver bitblast. (#5851)

This PR adds a new bit-blasting BV solver, which can be enabled via --bv-solver=bitblast. The new bit-blast solver can be used instead of the lazy BV solver and currently supports CaDiCaL and CryptoMiniSat as SAT back end.

HanielB committed:
[proof-new] Catch trivial cycles in SAT proof generation (#5853)

Evaluating the proof infrastructure in SMT-LIB has uncovered a rare
case (i.e., not previously in our regressions!!) in which we generate
a trivial cycle during SAT proof generation, which can lead to
problematic cycles when expanding MACRO_RESOLUTION steps. For example,
we can go from

       l1 v l2   ~l1 v l3   ~l2 v l3
(ok) ------------------------------
                  l3
in which l3 = l1 v l2, to

         l1 v l2   ~l1 v l3   ~l2 v l3
(bad) ------------------------------
               l3 v l3
              ---------
                 l3
This commit now catches the trivial cycle before it's generated.

ajreynol committed:
Eliminate equality query dependence on quantifiers engine (#5831)

This class will be renamed "RepSelector" on a future PR. It no longer needs to depend on quantifiers engine and can be initialized after the modules it depends on.

yoni206 committed:
Clarifying documentation of `--static-binary` (#5844)

The documentation regarding --static-binary is improved.

ajreynol committed:
Introduce quantifiers registry utility (#5829)

This is a simple module for determining which quantifiers module has ownership of quantified formulas.

This is work towards eliminating dependencies of quantifiers modules.

Note that quantifiers attributes module (which no longer has a dependency on QuantifiersEngine after this PR) will be embedded into this module in a later PR.

yoni206 committed:
Adding an option to optimize polite combination for datatypes (#5856)

This PR makes the optimization introduced in bbca987 optional.

ajreynol committed:
Miscellaneous cleaning in theory engine (#5854)

This statistic is unused, I believe it is leftover from deleted code.

aniemetz committed:
google test: expr: Migrate node_builder_black. (#5855)

aniemetz committed:
Remove obsolete include from node_black unit test. (#5862)

aniemetz committed:
google test: expr: Migrate node_manager_white. (#5864)

ajreynol committed:
Minor cleaning of quantifiers engine (#5858)

This member is currently unused.

aniemetz committed:
google test: expr: Migrate symbol_table_black. (#5870)

ajreynol committed:
Do not combine theories if theory engine needs check (#5861)

In rare cases, theory combination would be run when theory engine still needs check. This was limited to cases where the output channel is used but no lemmas were sent (TheoryEngine::needCheck() vs. d_lemmasAdded).

This led to problems in the theory of sets, which does not run a full effort check if theory engine needs check (since it knows it will be called to check again). However, running theory combination in these cases is not safe since theory of sets computes information pertaining to the care graph during its full effort check. Running theory combination otherwise would lead to theory of sets using stale data, leading to crashes due to terms not appearing in the equality engine.

This fixes #4124 (which appears not to trigger on master anyways currently).

This bug has also appeared on my sat relevancy development branch, hence fixing on master.

ajreynol committed:
Avoid spurious traversal of terms during preregistration (#5860)

This simplifies a few places in the code which unecessarily traverse terms during preregistration (which already traverses terms).

aniemetz committed:
google test: expr: Migrate node_traversal_black. (#5868)

aniemetz committed:
Use consistent names for fixtures in unit tests. (#5863)

ajreynol committed:
Remove support for inst closure (#5874)

This was a feature implemented for "Deciding Local Theory Extensions via E-matching" CAV 2015 that is not used anymore, and will be a burden to maintain further with potential changes to term database.

It also simplifies the TermDatabase::addTerm method (which changed indentation).

HanielB committed:
Fix dumping of assertions for monolithic preprocessing passes (#5866)

Previously we were unable to dump assertions in ProcessAssertions::apply for the definition-expansion, simplify and repeat-simplify passes.

aniemetz committed:
google test: expr: Migrate node_white. (#5869)

ajreynol committed:
Fix disequality between seq.unit terms (#5880)

This adds a missing inference for disequality between seq.unit terms, which was not handled previously.

Fixes #5666.

ajreynol committed:
Use quantifiers inference manager for lemma management (#5867)

Towards eliminating dependencies on quantifiers engine.

This replaces the custom implementation of lemma management in quantifiers engine with the standard one.

It makes a few minor changes to the standard inference manager classes to ensure the same behavior for quantifiers.

Note that some minor changes in behavior are introduced in this PR, such as being more consistent about caching/rewriting lemmas. This should not have any major impact.

aniemetz committed:
google test: expr: Migrate type_node_white. (#5872)

HanielB committed:
[quantifiers] Fix prenex computation (#5879)

Previously our prenex computation could generate quantifiers of the form forall x y y. F, which would lead to an assertion failure in getFreeVariablesScope, as it assumes that no shadowing occurs. This commit makes the prenex computation take a set rather than a vector, thus avoiding duplications of prenexed variables. It also changes mkForall to take a constant vector, since it does not modify the given vector.

Fixes #5693

aniemetz committed:
google test: expr: Migrate node_self_iterator_black. (#5865)

aniemetz committed:
google test: expr: Migrate type_cardinality_black. (#5871)

Commit:
syntax and documentation fixes

aniemetz committed:
google test: main: Migrate interactive_shell_black. (#5878)

aniemetz committed:
google test: expr: Migrate node_manager_black. (#5857)

ajreynol committed:
Make term database optionally SAT-context-dependent (#5877)

This makes the terms registered to the term database (those considered by E-matching) optionally stored in a SAT-context-dependent manner. The motivation is to have a more flexible/fine-grained set of terms considered by E-matching, e.g. if preregistration becomes lazier in the future.

This uncovered 2 issues:

The induction techniques in "conjecture generator" were using private interfaces, this PR removes the friend relaionship and cleans the code
The conflict-based instantiation module was accessing the signature tables for BOUND_VARIABLES when an operator of an APPLY_UF was a BOUND_VARIABLE. This is possible when options::ufHo is enabled. This makes conflict-based instantiation skip such terms.

ajreynol committed:
Eliminating dependencies from inst utils (#5882)

Towards eliminating dependence on QuantifierEngine from inst_match_trie.

mpreiner committed:
cmake: Make Python3 default and improve toml error messages. (#5884)

./configure.sh will now fail if Python3 is not installed on the system. Since Python2 is now deprecated the user has to explicitly enable it via --python2. This commit also removes the --python3 configure flag.

ajreynol committed:
Do not traverse quantifiers in term formula removal (#5859)

Our current policy for term formula removal leaves quantifier bodies unchanged. This simplifies the code based on this observation.

ajreynol committed:
Optimize get skolems method (#5876)

This method is used to determine which skolems have definitions according to term formula removal. This optimizes the method using attributes similar to how expr::getFreeVariables works.

It will be used heavily for SAT relevancy. We should also consider updating the justification heuristic to use this method.

ajreynol committed:
Remove track instantiations infrastructure (#5883)

This was used in the old method for unsat core tracking for instantiation lemmas, which will soon be subsumed.

This is also work towards eliminating the dependencies on quantifiers engine from Instantiate.

ajreynol committed:
Simplify method for inferring proxy lemmas in strings (#5789)

This PR simplifies our heuristic for inferring when an explanation for a strings lemma can be minimized based on proxy variable definitions. In particular, we do not turn solved equalities for proxy variables into substitutions. This aspect of the heuristic is incompatible with the new eager solver work, and moreover is buggy since substitutions should not be applied within string terms. The latter was leading the incorrect models on the 2 issues fixed by this PR.

Current results on the eager solver on SMT-LIB shows this change has very little performance impact.

Fixes #5692, fixes #5610.

ajreynol committed:
Fix open proof for factoring lemma (#5885)

We need to add to the current proof, regardless of whether we have used the factor skolem previously.

Fixes bugs found by @HanielB on SMT-LIB runs.

ajreynol committed:
Refactor term registration visitors (#5875)

This refactors the term registration visitors (PreRegisterVisitor and SharedTermsVisitor), which are responsible for calling Theory::preRegisterTerm and Theory::addSharedTerm.

The purpose of this PR is to resolve a subtle bug in the implementation of PreRegisterVisitor (see explanation below), and to improve performance in particular on the Facebook benchmarks (where preregistering terms is currently 25-35% of runtime on certain challenge benchmarks).

This PR makes it so that that SharedTermsVisitor now subsumes PreRegisterVisitor and should be run when sharing is enabled only. Previously, we ran both PreRegisterVisitor and SharedTermsVisitor, and moreover the former misclassified when a literal had multiple theories since the d_theories field of PreRegisterVisitor is never reset. This meant we would consider a literal to have multiple theories as soon as we saw any literal that had multipleTheories. After this PR, we run SharedTermsVisitor only which also handles calling preRegisterTerm, since it has the ability to anyways (it computes the same information as PreRegisterVisitor regarding which theories care about a term).

The refactoring merges the blocks of code that were copy and pasted in term_registration_visitor.cpp and makes them more optimal, by reducing our calls to Theory::theoryOf.

FYI @barrettcw

ajreynol committed:
Simplify interface for preprocessor (#5890)

This does a minor simplification to the interface for preprocessor. It has a postProcess call that is called after another unrelated call at the SmtSolver level (notifyPreprocessedAssertions) is made. This makes it so that Preprocessor::postProcess is inlined as the last step of process.

ajreynol committed:
Fix spurious assertion failure in regexp normalization (#5852)

Fixes #5816.

nafur committed:
Merge InferenceIds into one enum (#5892)

This PR is the first step in consolidating the various variants of Inferences, InferenceIds and InferenceManagers of the theories.
It merges all InferencedIds into one global enum and makes all theories use them.

HanielB committed:
[proof-new] Adding a proof-producing ensure literal method (#5889)

The ensureLiteral method in CnfStream may apply CNF conversion to the given literal (for example if it's an IFF), which needs to be recorded in the proof. This commit adds a proof-producing ensureLiteral to ProofCnfStream, which is called by the prop engine if proofs are enabled.

This commit also simplifies the interfaces on ensureLit and convertAtom by removing the preRegistration flag, which was never used.

Commit:
Add InferenceId member to TheoryInference, adapt all derived classes. (#5894)

This PR adds a new InferenceId member to the TheoryInference class.
All classes derived from TheoryInference are adapted accordingly.

Commit:
Make most methods of TheoryInferenceManager expect an InferenceId. (#5897)

This PR makes most methods of the TheoryInferenceManager expect an InferenceId.
All classes that inherit from TheoryInferenceManager are adapted accordingly and InferenceIds are passed everywhere.
In some cases, there are no appropriate InferenceIds yet. We use InferenceId::UNKNOWN for the time being and introduce proper values in future PRs.
The InferenceIds are not yet used, but will be used for statistics and debug output.

aniemetz committed:
google test: parser: Migrate parser_black. (#5886)

ajreynol committed:
(proof-new) Option to not automatically consider symmetry in CDProof (#5895)

There are compelling use cases not to automatically introduce SYMM steps in CDProof, e.g. in CDProofs used within ProofNodeUpdater for external conversions.

ajreynol committed:
Simplify and fix decision engine's handling of skolem definitions (#5888)

This PR changes the front end of prop engine to distinguish input formulas from skolem definitions, which is required by the decision engine. This PR breaks the dependency of PropEngine on the AssertionsPipeline, as now the determining of whether an input formula is a skolem definition is handled externally, in SmtSolver.

With this PR, we should not be required to satisfy skolem definitions that are not relevant based on the techniques already implemented in the decision engine. Currently, we are not distinguishing input formulas from skolem definitions properly, meaning we assert more literals than we need to.

This also simplifies related interfaces within decision engine.

We should test this PR with --decision=justification on SMT-LIB.

mpreiner committed:
Properly set up equality engine for BV bitblast solver. (#5905)

Theory BV now sets up the default equality engine for BV solvers that do not use their own equality engine like e.g. the BV bitblast solver. This commit also adds the missing equality engine pieces to the BV bitblast solver (getEqualityStatus, explain).

ajreynol committed:
Moving methods from quantifiers engine to quantifiers state (#5881)

Towards eliminating dependencies on quantifiers engine.

nafur committed:
Remove now obsolete sendLemmas and inferences stat from arith::nl (#5903)

This PR removes some obsolete code from the nonlinear solver.
The statistics will soon be replaced by a generic statistic in the theory inference manager.

aniemetz committed:
google test: parser: Migrate parser_builder_black. (#5896)

mpreiner committed:
Add bit-level propagation support to BV bitblast solver. (#5906)

This commit adds support for bit-level propagation for the BV bitblast solver to quickly detect conflicts on effort levels != FULL. Bit-level propagation for the bitblast solver is by default disabled for now. Further, bit-blasting of facts is now handled more lazily with a bit-blast queue.

Commit:
Add new IntegralHistogramStat (#5898)

This PR adds a new statistics class that improves on HistogramStat if we know that the type is integral (also supporting enums).
Instead of using a comparably slow std::map like HistogramStat, IntegralHistogramStat uses a std::vector that is resized appropriately.
The integral values are simply cast to std::int64_t and used as indices into the vector.

Commit:
Add InferenceId to buffered inference manager (#5911)

This PR collects the InferenceId in the InferenceManagerBuffered class.

Commit:
Add InferenceIds for theory of arrays (#5910)

This PR introduces new InferenceIds for the theory of arrays and uses them instead of InferenceId::UNKNOWN.

Commit:
TheoryIds for UF theory. (#5901)

This PR introduces new InferenceId for the uf theory and uses them instead of InferenceId::UNKNOWN.

Commit:
Use InferenceId in sep theory. (#5912)

This PR uses the new InferenceIds in the separation logic theory.

ajreynol committed:
Compute fact or lemma in datatypes prior to buffering (#5914)

This is necessary for the planned refactoring of TheoryInference::process. This forces datatypes to decide lemma vs. fact prior to buffering inferences.

ajreynol committed:
Move methods from term util to quantifiers registry (#5916)

Towards eliminating dependencies on quantifiers engine, and eliminating the TermUtil class.

Note that QuantifiersModule had to be moved to its own file to avoid circular include dependencies.

ajreynol committed:
Move first order model for full model check to own file (#5918)

This moves the derived model class used in finite model finding to its own file, in the src/theory/quantifiers/fmf directory.

Updates the code to meet guidelines, no behavior changes.

ajreynol committed:
Eliminate non-static members in term util (#5919)

This makes it so that TermUtil is now a collection of static methods. Further refactoring will make this a standalone file of utility methods.

This breaks all dependencies on the TermUtil object in QuantifiersEngine. It also starts breaking some of the depenendencies on quantifiers engine in sygus.

ajreynol committed:
Document UF inferences (#5917)

Document UF entries of InferenceId enum.

Commit:
New InferenceIds for BV theory (#5909)

This PR introduces new InferenceId for the BV theory and uses them instead of InferenceId::UNKNOWN.

Commit:
Add InferenceIds for sets theory. (#5900)

This PR introduces new InferenceId for the theory of sets and uses them instead of InferenceId::UNKNOWN.

Commit:
Add statistic for InferenceId to TheoryInferenceManager. (#5913)

This PR uses the IntegralHistogramStat to obtain statistics about the sent inferences within the TheoryInferenceManager. All theories are adapted to provide a proper name (prefix) for the name of the statistic.

aniemetz committed:
Add Gereon to AUTHORS list. (#5930)

Commit:
Cleanup of inferences in arithmetic theory (#5927)

This PR cleans up several issues in the arithmetic theory mostly related to its usage of InferenceId and the TheoryInferenceManager:

    remove the ArithLemma class and uses SimpleTheoryLemma instead
    only use NlLemma if we actually need it
    use proper InferenceIds everywhere
    remove unused code in the nonlinear extension

ajreynol committed:
Remove string stat for inferences (#5932)

This is now subsumed by the general stat in TheoryInferenceManager

ajreynol committed:
Fix rewrite for contains over replace (#5924)

Fixes model soundness issue (fixes #5915).

ajreynol committed:
Refactoring theory inference process (#5920)

This PR refactors TheoryInference so that it is not responsible for calling back into InferenceManager, instead it sets data so that InferenceManagerBuffered has enough information to do this itself. It also makes the decision of whether to cache lemmas in theory inference manager a global choice per-theory instead of per-lemma.

ajreynol committed:
Fill in missing inference ids in datatypes theory (#5931)

Also updates its inference manager to not track stats since the standard ones are now used.

This also sets up some dependencies in the sygus extension which will be used to implement InferenceId for the sygus extension, to be done on a separate PR.

ajreynol committed:
Simplify interface to instantiate (#5926)

Does not support InstMatch interfaces anymore, which are spurious.

ajreynol committed:
Fix datatypes inference manager when proofs are enabled (#5937)

Accidentally was not sending lemmas in one interface when proofs are enabled due to recent refactoring.

nafur committed:
add pruneRedundantIntervals (#5950)

Adds a simple helper for CAD to prune redundant intervals. It is just a wrapper for cleanIntervals right now, but will be responsible to making sure the CAD proof is pruned as well.

nafur committed:
Cleanup in transcendental solver, add ApproximationBounds struct. (#5945)

This PR merges some cleanup in the transcendental solver from proof-new.
It adds a new struct ApproximationBounds that replaces an opaque std::vector and does some general refactoring in the TaylorGenerator class, removing dead code and using fixed-width integers.

nafur committed:
(proof-new) Add proofs for sine lemmas in the transcendental solver (#5952)

This PR adds proofs for the lemmas related to the sine function in the transcendental solver.
It introduces several new proof rules with corresponding proof checkers and produces proofs in the sine solver.

nafur committed:
(proof-new) Add new arithmetic kind INDEXED_ROOT_PREDICATE. (#5949)

This PR introduces a new arithmetic kind for indexed root predicates.
An indexed root predicate compares a real variable to the k'th root of a given polynomial as follows:
Let IRP_k(x ~ 0, p) an indexed root predicate with k a non-negative number, x some real variable, ~ an arithmetic relation (e.g. =, <, ...), and p a polynomial over x (and possibly other variables).
If p contains variables apart from x, we can only evaluate the expression over a suitable assignment for at least these variables.
The evaluation of this expression is equivalent to computing the k'th real root of p in x (with all other variables evaluated over a given assignment) and comparing this real root to zero (according to the relation symbol ~).

Note that we currently do not intend to use this structure for solving, but require it for representing and printing CAD proofs.

nafur committed:
Add the LazyTreeProofGenerator. (#5948)

This PR adds a new proof utility to construct tree-shaped proofs in a lazy fashion.
The LazyTreeProofGenerator is currently intended to be used for CAD proofs, where we need to construct proofs that consist of nested case-splits, but the exact split (in a form suitable for proof construction) is only known when the whole subtree is finished.
We thus store the proof in a tree structure similar to proof nodes, and transform the whole proof to a proper proof node once all data is available.

aniemetz committed:
google test: preprocessing: Migrate pass_foreign_theory_rewrite_white. (#5936)

HanielB committed:
[proof-new] Optionally print conclusion in the AST proof (#5954)

Adds an option to optionally print conclusion in the AST proof.

ajreynol committed:
(proof-new) Option to automatically add SYMM steps during proof node update. (#5939)

Required for work on external proof conversions.

ajreynol committed:
Require length-in-conclusion form for strings inferences (#5953)

Previously, it was optional whether to put length constraints in conclusion for deq string inferences. However, due to optimizations in skolem caching, it is now required to guard these length constraints.

It furthermore changes the policy to not ignore the lengths for the registered skolem, since it may be shared elsewhere.

Fixes #5940.

Notice that proof-new already requires this option to be enabled when proofs are enabled. Hence, this will simplify proof-new when merged.

nafur committed:
(proof-new) Add proofs for exponential functions (#5956)

This PR adds proofs for lemmas concerned with the exponential function. If also adds the necessary proof rules and corresponding proof checker.

ajreynol committed:
(proof-new) Change proof-new option to proof (#5955)

Also moves several proof-specific options to proof_options.

ajreynol committed:
Move quantifiers attributes to quantifiers registry (#5929)

This moves the utility class beneath quantifiers registry.

ajreynol committed:
Eliminate raw use of output channel and valuation in quantifiers (#5933)

This makes all lemmas in quantifiers sent through the inference manager. It begins adding InferenceId values for some of these calls. All uses of Valuation are replaced by calls to QuantifiersState.

aniemetz committed:
google test: preprocessing: Migrate pass_bv_gauss_white. (#5935)

ajreynol committed:
Explanation of failure for instantiate, use in enumerative instantiation (#5951)

This makes enumerative instantiation generalize the failures in Instantiate::addInstantiate and increment its enumeration accordingly.

This leads to (+104-6) using enumerative instantiation only on SMT-LIB quantified benchmarks, 60 second timeout.

This is in preparation for further improvements to enumerative instantiation.

nafur committed:
Add trans secant proofs. (#5957)

This PR adds proofs for secant lemmas in the transcendental lemmas for both exponential and sine functions.
It also adds proof rules and corresponding proof checkers.

HanielB committed:
[proof-new] Fix handling of removable clauses in proof cnf stream (#5961)

Previously the removable information was not being communicated from the proof cnf stream to the cnf stream, which is the one that actually uses this when asserting clauses into the SAT solver. This commit fixes this by having the proof cnf stream directly use the cnf stream d_removable attribute.

nafur committed:
(proof-new) Add proof generator for CAD solver (#5964)

This PR adds a proof generator for the CAD solver, including two new proof rules. The code is not yet used, but will be integrated into the CAD solver itself in another PR.

nafur committed:
Add proof for monomial bounds check (#5965)

This PR adds proofs for a nonlinear refinement lemma that deals with multiplication of inequalities with some term. This lemma is split into two proof rules for positive or negative factors.

HanielB committed:
[proof-new] Fix dangling pointer in SAT proof generation (#5963)

When a clause is being explained, the negation of each of its literals, other than the one it propagates, needs to be explained. This process may lead to the creation of new clauses in the SAT solver (because if a literal being explained was propagated and an explanation was not yet given, it will then be computed and added). This may lead to changes in the memory where clauses are, which may break the reference to the original clause being explained. To avoid this issue we store the literals in the reason before we start explaining their negations. We then iterate over them rather than over the clause, as before.

yoni206 committed:
Merge branch 'master' into optimization_subsolver

nafur committed:
Add proof for mult sign lemma (#5966)

This PR adds the proof for a nonlinear refinement lemma that infers the sign of a monomial from the sign of the variables.

aniemetz committed:
google test: theory: Migrate strings_rewriter_white. (#5976)

aniemetz committed:
google test: theory: Migrate regexp_operation_black. (#5974)

mpreiner committed:
Switch to C++17. (#5959)

Co-authored-by: Gereon Kremer <nafur42@gmail.com>

ajreynol committed:
Add interface to TheoryState for sort inference and facts (#5967)

This eliminates the need for direct references to TheoryEngine from quantifiers and UF+cardinality.

This PR also eliminates an unnecessary reference to TheoryEngine in TheoryModelBuilder and breaks a few more dependencies in quantifiers modules.

ajreynol committed:
Add state and inference manager to inst match generator (#5968)

In preparation for refactoring E-matching to not pass QuantifiersEngine pointer to its utilities.

aniemetz committed:
google test: prop: Migrate cnf_stream_white. (#5971)

aniemetz committed:
google test: theory: Migrate evaluator_white. (#5972)

aniemetz committed:
google test: theory: Migrate theory_arith_white. (#5977)

aniemetz committed:
google test: theory: Migrate theory_bags_rewriter_white. (#5979)

aniemetz committed:
google test: theory: Migrate logic_info_white. (#5973)

aniemetz committed:
google test: theory: Migrate theory_bags_normal_form_white. (#5978)

aniemetz committed:
google test: theory: Migrate sequences_rewriter_white. (#5975)

nafur committed:
(proof-new) Add proofs for CAD solver (#5981)

This PR adds proofs for the CAD solver, based on the proof generator from the previous PR.
Note that the level of detail of these CAD proofs is significantly higher than for other proofs. Making these proofs more fine-grained and maybe at some point accessible to proof checkers is probably still quite a bit of work.
Thus, the CAD proof rules are both trusted rules for now.

nafur committed:
Ensure static-learning adds rewritten assertions. (#5982)

The static-learning preprocessing sometimes added non-rewritten assertions, despite being used in a part of the preprocessor that assumes all assertions to be rewritten. This may then break other passes further down, in the case of #5729 the non-clausal simplification which explicitly asserts that assertions are rewritten. This PR rewrites the respective assertion properly in the static-learning pass.
Fixes #5729.

aniemetz committed:
google test: theory: Migrate theory_black. (#5985)

aniemetz committed:
google test: theory: Migrate theory_engine_white. (#5988)

aniemetz committed:
google test: theory: Migrate theory_bags_type_rules_white. (#5984)

mpreiner committed:
Enable -Werror. (#5969)

aniemetz committed:
google test: printer: Migrate smt2_printer_black. (#5970)

aniemetz committed:
google test: theory: Migrate theory_bv_rewriter_white. (#5986)

aniemetz committed:
google test: theory: Migrate theory_bv_white. (#5987)

nafur committed:
Add regression. (#5994)

This PR adds the test case from #5187 as a regression.
Fixes #5187.

aniemetz committed:
google test: Merge Node(Manager) fixtures. (#5998)

yoni206 committed:
Datatypes lemmas: share only external types (#5997)

Forcing lemmas in datatypes used to be done only for external types.
This was changed to consider all types, which is not needed.
This PR brings back the restriction to external types.

ajreynol committed:
Move slow regressions to regress1 (#5999)

Moves regressions taking >4 seconds (summing all configs) in debug to regress1.

nafur committed:
Move (optional) rewrite from TrustSubstitutionMap to SubstitutionMap. (#5992)

This PR adds optional rewriting to the SubstitutionMap class. Before, only the new TrustSubstitutionMap added optional rewriting, leading to unexpected inconsistencies between the two. In particular, cases exist where the substitution and the rewriting cancel each other (see #5943).
This PR moves the optional rewriting from TrustSubstitutionMap into SubstitutionMap. While the former enables it by default, it is disabled by default for the latter and thus does not change current behavior.
We now use this new option in an assertion in the non-clausal simplification.
Fixes #5943.

ajreynol committed:
(proof-new) Fix regular expression unfolding under substitution (#5958)

This case was previously unhandled and exercised by a recently added regression.

nafur committed:
Store Node instead of TNode (#5993)


The justification heuristic stores a "copy" of assertions as TNode. As witnessed by #5938, these TNodes may invalid.
This PR changes this to store Nodes instead.
Fixes #5938.

ajreynol committed:
Optionally permit creation of non-flat function types (#6010)

This is required for creating the representation of closues in LFSC, which are of the form ((forall x T) P) where notice that forall has non-flat function type (-> Int Sort (-> Bool Bool)).

ajreynol committed:
Minor improvement and fix to smt2 printer (#6009)

This permits access to the static method string smtKindString(Kind k, Variant v) which is required for LFSC proof conversion. It also makes a fix to how a string kind is printed.

aniemetz committed:
google test: util: Migrate assert_white. (#6011)

aniemetz committed:
google test: util: Migrate binary_heap_black. (#6012)

mpreiner committed:
Fix -Werror issues with clang and use clang for debug-cln build. (#6004)

This fixes some issues that break the nightly ASAN builds with clang.

aniemetz committed:
google test: theory: Migrate theory_sets_type_enumerator_white. (#6000)

aniemetz committed:
google test: theory: Migrate theory_sets_type_rules_white. (#6001)

HanielB committed:
Some formatting and better tracing in prop engine (#6022)

Miscellaneous changes from proof-new.

aniemetz committed:
google test: theory: Migrate type_enumerator_white. (#6007)

aniemetz committed:
google test: theory: Migrate theory_white. (#6006)

This moves test utils for theory tests to test_smt.h and consolidates
two implementations of dummy theories into one.

aniemetz committed:
google test: util: Migrate cardinality_black. (#6016)

aniemetz committed:
google test: util: Migrate check_white. (#6017)

aniemetz committed:
google test: util: Migrate exception_black. (#6020)

aniemetz committed:
google test: theory: Migrate theory_strings_word_white. (#6003)

aniemetz committed:
google test: theory: Migrate theory_strings_skolem_cache_black. (#6002)

aniemetz committed:
google test: util: Migrate integer_white. (#6024)

aniemetz committed:
google test: util: Migrate array_store_all_white. (#6008)

aniemetz committed:
google test: util: Migrate boolean_simplification_black. (#6014)

aniemetz committed:
google test: util: Migrate configuration_black. (#6018)

aniemetz committed:
google test: util: Migrate datatype_black. (#6019)

aniemetz committed:
google test: util: Migrate integer_black. (#6023)

aniemetz committed:
google test: util: Migrate output_black. (#6025)

aniemetz committed:
google test: util: Migrate rational_black. (#6026)

aniemetz committed:
google test: util: Migrate rational_white. (#6027)

aniemetz committed:
google test: util: Migrate real_algebraic_number_black. (#6028)

aniemetz committed:
google test: util: Migrate stats_black. (#6029)

mpreiner committed:
Make -Werror optional but enable it for CI. (#6032)

nafur committed:
Refactor collection of debug and trace tags (#5996)

We have a mechanism to collect all debug and trace tags used throughout the code base to allow checking for valid tags.
This mechanism relies on a collection of more or less readable shell scripts.
#5921 hinted to a problem with the current setup, as it passes all source files via command line.
This PR refactors this setup so that the scripts collect the files internally, and only the base directory is passed on the command line.
As I was touching this code anyway, I ported everything to python and combined it into a single script, in the hope to make it more maintainable.
Fixes #5921.

aniemetz committed:
google test: theory: Migrate theory_quantifiers_bv_instantiator_white. (#5989)

aniemetz committed:
google test: util: Migrate bitvector_black. (#6015)

aniemetz committed:
google test: theory: Migrate theory_quantifiers_bv_inverter_white. (#5991)

aniemetz committed:
google test: util: Migrate floatingpoint_black. (#6021)

nafur committed:
Clean up includes to reduce compile times (#6031)

This PR cleans up a ton of includes, based on the suggestions of iwyu.
Mostly, it removes includes from header files in favor of forward declarations and adds includes to source files.

ajreynol committed:
Introduce quantifiers term registry (#5983)

This groups utilities related to ground terms into TermRegistry which will be passed to quantifier modules.

mpreiner committed:
Fix nightly errors. (#6034)

Fixes warnings with CVC4_FALLTHROUGH and -Werror for debug/production with gcc/clang. Clang detects that a CVC4_FALLTHROUGH after an Assert(false); is unreachable and issues a warning, while gcc issues a warning about an implicit fall-through if CVC4_FALLTHROUGH is not present.

mpreiner committed:
Add aarch64 (ARM64) cross-compile support. (#6033)

This commit adds support for cross-compiling for aarch64 platforms and simplifies cross-compilation handling for Windows. The configure script now automatically downloads and cross-compiles the required dependencies ANTLR3 and GMP when passing option --arm64 or --win64.

Fixes #1479 #5769.

Commit:
removed opt_sat_partial and added small naming changes

Commit:
Merge branch 'optimization_subsolver' of https://github.com/mcjuneho/CVC4 into optimization_subsolver

aniemetz committed:
Remove obsolete dependency on CxxTest. (#6038)

nafur committed:
Improve handling of utf8 encoded inputs (#5694)

This PR fixes an issue where utf8 encoded inputs are incorrectly parsed into CVC4::String. We now use std::mbtowc to first turn the char sequence from the std::string input into a std::wstring and then process this std::wstring one charactor (wchar_t) at a time.
Fixes #5673

mpreiner committed:
Remove non-ASCII characters from source files. (#6039)

Make collect_tags.py more robust for non-ASCII characters.

mudathirmahgoub committed:
Add tuple projection operator (#5904)

This PR adds tuple projection operator to the theory of data types.
It Also adds helper functions for selecting elements from a tuple.

abdoo8080 committed:
Remove uses of SExpr class. (#6035)

This PR is a step towards removing SExpr class. It replaces SExpr with std::string for set-info and set-option commands.

nafur committed:
More cleanup of includes to reduce compilation times (#6037)

Similar to #6031, this PR implements suggestions from iwyu to reduce the number of includes in header files by introducing forward declarations and moving includes to source files.

ajreynol committed:
(proof-new) Simplifications to arithmetic operator elimination and preprocessing (#6040)

Due to refactoring in theory preprocessor, which does fixed point preprocessing on created lemmas, several things can be simplified in arithmetic operator elimination.

This is required for further simplification to witness forms in the internal proof calculus.

nafur committed:
Remove obsolete gcc check. (#6041)

This PR removes a cmake check for gcc 4.5.1. As this version is more than a decade old and would not work anyway, as it does not fully support C++11, not to mention C++17.

aniemetz committed:
New C++ API: Clean up usage of internal types in Op. (#6045)

This disables the temporarily available internals of Op.

aniemetz committed:
New C++ API: Clean up usage of internal Result. (#6043)

This disables the temporarily available internals of Result.
It further changes the interface for getUnknownExplanation, which now
returns an enum value instead of a string.

aniemetz committed:
New C++ API: Clean up usage of internal Type/TypeNodes. (#6044)

This disables the temporarily available internals of Sort.

aniemetz committed:
context_black: Clean up classes. (#6046)

This cleans up the MyContext* classes defined for the tests according to
the code style guidelines. It further converts non-fixed width integer
types to fixed-width types. This was missed in #5587.

aniemetz committed:
Fix nightlies. (#6048)

nafur committed:
Ignore isInConflict when adding conflicts (#5995)

Right now, the inference manager infrastructure drops conflicts (and literal propagations) if the solver already is in a conflict.
This PR removes this behavior. The current setup in linear arithmetic requires adding conflicts, even when already in conflict, and experiments showed a small but beneficial effect of this change.

Commit:
updated to master and 2nd attempt at fixing merge conflicts

aniemetz committed:
Fix nightlies. (#6052)

nafur committed:
Add cmake scripts for iwyu targets. (#6042)

This PR adds some utility targets that simplify the usage of iwyu (include-what-you-use) on our code base.

nafur committed:
Add proper define for libpoly usage (#6050)

When adding libpoly, we forgot to add a proper define to cvc4autoconfig and included real_algebraic_number.h everywhere to get this define. This PR introduces the CVC4_POLY_IMP define and removes all obsolete includes to real_algebraic_number.h.

aniemetz committed:
New C++ Api: Clean up usage of internal types in Term. (#6054)

This disables the temporarily available internals of Term.
Note: getExpr() is still available and will be disabled when the API is
      fully converted to Nodes.

Commit:
Merge branch 'master' into optimization_subsolver

Commit:
change

aniemetz committed:
Add initial bit-blaster for proof logging. (#6053)

Co-authored-by: Mathias Preiner <mathias.preiner@gmail.com>

Commit:
brought unit test up to date with google test

Commit:
reverted ci.yml file

aniemetz committed:
New C++ API: Clean up usage of interal datatype classes. (#6055)

This disables the temporarily available internals of datatype classes.

aniemetz committed:
google test: Remove dependency on ExprManager in type_cardinality_black. (#6061)

Commit:
Merge branch 'master' into optimization_subsolver

nafur committed:
Reimplement time limit mechanism for windows (#6049)

As noted in #5034, --tlimit is not working properly on windows. It turns out that the timer mechanism provided by the windows API are not suitable for our use case. Thus, this PR implements a generic std::thread-based timer mechanism which is used whenever the POSIX timers (setitimer) are not available. It also adds some documentation on the timer options and the reasons we ended up with this.
Fixes #5034.

aniemetz committed:
google test: Remove obsolete Expr test fixtures. (#6060)

yoni206 committed:
Merge branch 'master' into optimization_subsolver

Commit:
small style fixes to unit test

Commit:
Merge branch 'optimization_subsolver' of https://github.com/mcjuneho/CVC4 into optimization_subsolver

Commit:
whoops deleted the class members

Commit:
changed to TestSmtNoFinishInit
